### PR TITLE
feat: Exponential Search

### PR DESCRIPTION
**Pull request by faridevnz**
#### Description of Change
- added Exponential Search Algorithm

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: implementation of Exponential Search

### CODE REVIEW

File: searching/exponential_search.c [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    if ( arr[middle_index] == n ) return middle_index;
+    // recursion
+    if ( arr[middle_index] > n ) return binary_search(arr, l_index, middle_index-1, n); // left
+    return binary_search(arr, middle_index+1, r_index, n); // right
+}
+
+/**
+ *  Function: main
+ *  --------------
+ */
```
Comment (Panquesito7) [inline searching/exponential_search.c:63]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
```

File: searching/exponential_search.c [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+int main(void) 
+{
+    // perform the tests
+    test();
```
Comment (Panquesito7) [inline searching/exponential_search.c:67]:
```suggestion
    test();  // run self-test implementations
```

File: searching/exponential_search.c [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    return 0;
+}
+
+/**
+ *  Function: test
+ *  --------------
+ */
```
Comment (Panquesito7) [inline searching/exponential_search.c:74]:
```suggestion
/**
 * @brief Self-test implementations
 * @returns void
 */
```

File: searching/exponential_search.c [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+void test(void)
```
Comment (Panquesito7) [inline searching/exponential_search.c:75]:
```suggestion
static void test()
```

File: searching/exponential_search.c [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    // recursion
+    if ( arr[middle_index] > n ) return binary_search(arr, l_index, middle_index-1, n); // left
+    return binary_search(arr, middle_index+1, r_index, n); // right
+}
+
+/**
+ *  Function: main
+ *  --------------
+ */
+int main(void)
```
Comment (Panquesito7) [inline searching/exponential_search.c:64]:
```suggestion
int main()
```

File: searching/exponential_search.c [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+ *  \param n:           element to find in the array (arr)
+ * 
+ *  \returns:           the index of the element (n) in the array (arr)
+ *  \returns:           -1 if the element wasn't found
+ */
+int exponential_search(const int* arr, const int length, const int n) 
+{
+    // find the upperbound
+    int upper_bound = 1;
+    while ( upper_bound <= length && arr[upper_bound] < n ) upper_bound = upper_bound * 2;
```
Comment (Panquesito7) [inline searching/exponential_search.c:29]:
```suggestion
    while ( upper_bound <= length && arr[upper_bound] < n ) { upper_bound = upper_bound * 2; }
```

File: searching/exponential_search.c [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    // calculate the range ( between lower_boud and upper_bound )
+    int lower_bound = upper_bound/2;
+    if ( upper_bound > length ) upper_bound = length;
```
Comment (Panquesito7) [inline searching/exponential_search.c:32]:
```suggestion
    if ( upper_bound > length ) { upper_bound = length; }
```

File: searching/exponential_search.c [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+ * 
+ *  \returns:           the index of the element (n) in the array (arr)
+ *  \returns:           -1 if the n element wasn't found
+ */
+int binary_search(const int* arr, const int l_index, const int r_index, const int n) 
+{
+    // calculate the middle index of the array
+    int middle_index = l_index + ( r_index - l_index ) / 2;
+    // base cases
+    if ( l_index > r_index ) return -1;
```
Comment (Panquesito7) [inline searching/exponential_search.c:53]:
```suggestion
    if ( l_index > r_index ) { return -1; }
```

File: searching/exponential_search.c [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    if ( arr[middle_index] == n ) return middle_index;
```
Comment (Panquesito7) [inline searching/exponential_search.c:54]:
```suggestion
    if ( arr[middle_index] == n ) { return middle_index; }
```

File: searching/exponential_search.c [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    // recursion
+    if ( arr[middle_index] > n ) return binary_search(arr, l_index, middle_index-1, n); // left
```
Comment (Panquesito7) [inline searching/exponential_search.c:56]:
```suggestion
    if ( arr[middle_index] > n ) { return binary_search(arr, l_index, middle_index-1, n); } // left
```

File: searching/exponential_search.c [hunk 11]
```diff
@@ -0,0 +1,101 @@
+/**
+ *  \file
+ *  \link Algorithm explanation:
+ *  https://github.com/TheAlgorithms/Algorithms-Explanation/blob/master/en/Search%20Algorithms/Exponential%20Search.md
```
Comment (Panquesito7) [inline searching/exponential_search.c:4]:
```suggestion
 *  \@brief [Exponential Search](https://github.com/TheAlgorithms/Algorithms-Explanation/blob/master/en/Search%20Algorithms/Exponential%20Search.md) implementation
```

File: searching/exponential_search.c [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+ *  \author [Alessio Farinelli] (https://github.com/faridevnz)
```
Comment (Panquesito7) [inline searching/exponential_search.c:5]:
```suggestion
 *  \author [Alessio Farinelli](https://github.com/faridevnz)
```

File: searching/exponential_search.c [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+ */
+#include <assert.h>
+
+#define ELEMENT -10
+
+int binary_search(const int* arr, const int l_index, const int r_index, const int n);
+int exponential_search(const int* arr, const int length, const int n);
+void test(void);
```
Comment (Panquesito7) [inline searching/exponential_search.c:13]:
```suggestion
static void test();
```

File: searching/exponential_search.c [hunk 14]
```diff
@@ -0,0 +1,101 @@
+/**
+ *  \file
+ *  \link Algorithm explanation:
+ *  https://github.com/TheAlgorithms/Algorithms-Explanation/blob/master/en/Search%20Algorithms/Exponential%20Search.md
+ *  \author [Alessio Farinelli] (https://github.com/faridevnz)
+ */
+#include <assert.h>
```
Comment (Panquesito7) [inline searching/exponential_search.c:7]:
```suggestion
#include <assert.h>   /// for assert
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
👍 Good work!
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/faridevnz/C/actions

File: searching/exponential_search.c [hunk 15]
```diff
@@ -0,0 +1,99 @@
+/**
+ *  \file
+ *  \@brief [Exponential Search](https://github.com/TheAlgorithms/Algorithms-Explanation/blob/master/en/Search%20Algorithms/Exponential%20Search.md)
```
Comment (Panquesito7) [inline searching/exponential_search.c:3]:
```suggestion
 *  \brief [Exponential Search](https://github.com/TheAlgorithms/Algorithms-Explanation/blob/master/en/Search%20Algorithms/Exponential%20Search.md)
```

File: searching/exponential_search.c [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+#define ELEMENT -10
+
+int binary_search(const int* arr, const int l_index, const int r_index, const int n);
+int exponential_search(const int* arr, const int length, const int n);
+static void test();
+
+/**
+ *  Function: exponential_search
+ *  ----------------------------
+ *  \param arr:         array where search the element
```
Comment (Panquesito7) [inline searching/exponential_search.c:17]:
```suggestion
 *  \param arr array where search the element
```

File: searching/exponential_search.c [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+ *  \param length:      the total length of the given array (arr)
+ *  \param n:           element to find in the array (arr)
+ * 
+ *  \returns:           the index of the element (n) in the array (arr)
+ *  \returns:           -1 if the element wasn't found
```
Comment (Panquesito7) [inline searching/exponential_search.c:22]:
```suggestion
 *  \param length the total length of the given array (arr)
 *  \param n the element to find in the array (arr)
 * 
 *  \returns the index of the element (n) in the array (arr)
 *  \returns -1 if the element wasn't found
```

File: searching/exponential_search.c [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+/**
+ *  Function: binary_search
+ *  ----------------------------
+ *  \param arr:         array where search the element
+ *  \param l_index:     start index of the array (arr) to apply the algorithm
+ *  \param r_index:     end index of the array (arr) to apply the algorithm
+ *  \param n:           element to find in the array (arr)
+ * 
+ *  \returns:           the index of the element (n) in the array (arr)
+ *  \returns:           -1 if the n element wasn't found
```
Comment (Panquesito7) [inline searching/exponential_search.c:45]:
```suggestion
 *  \param arr array where search the element
 *  \param l_index start index of the array (arr) to apply the algorithm
 *  \param r_index end index of the array (arr) to apply the algorithm
 *  \param n the element to find in the array (arr)
 * 
 *  \returns the index of the element (n) in the array (arr)
 *  \returns -1 if the n element wasn't found
```

File: searching/exponential_search.c [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+    while ( upper_bound <= length && arr[upper_bound] < n ) { upper_bound = upper_bound * 2; }
+    // calculate the range ( between lower_boud and upper_bound )
+    int lower_bound = upper_bound/2;
+    if ( upper_bound > length ) { upper_bound = length; }
+    // apply the binary search in the range
+    return binary_search(arr, lower_bound, upper_bound, n);
+}
+
+/**
+ *  Function: binary_search
```
Comment (Panquesito7) [inline searching/exponential_search.c:37]: Provide a brief description of what the function does or is for. For long/detailed description, use `\details`.

File: searching/exponential_search.c [hunk 20]
```diff
@@ -0,0 +1,99 @@
+/**
+ *  \file
+ *  \@brief [Exponential Search](https://github.com/TheAlgorithms/Algorithms-Explanation/blob/master/en/Search%20Algorithms/Exponential%20Search.md)
+ *  \author [Alessio Farinelli](https://github.com/faridevnz)
+ */
+#include <assert.h>  /// for assert
+
+#define ELEMENT -10
+
+int binary_search(const int* arr, const int l_index, const int r_index, const int n);
```
Comment (Panquesito7) [inline searching/exponential_search.c:10]: Consider using `uint64_t`, `uint32_t`, `uint16_t`, or even `uint8_t` for non-negative values. For negative values, use `int64_t` or their appropriate size ([reference](https://en.cppreference.com/w/c/types/integer)). Requires the `inttypes.h` header. Check out other parts of the code, too. 🙂

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: searching/exponential_search.c [hunk 21]
```diff
@@ -0,0 +1,104 @@
+/**
+ *  \file
+ *  \brief [Exponential Search](https://github.com/TheAlgorithms/Algorithms-Explanation/blob/master/en/Search%20Algorithms/Exponential%20Search.md)
+ *  \author [Alessio Farinelli](https://github.com/faridevnz)
+ */
+#include <assert.h>  // for assert
+#include <inttypes.h>  // for int64_t, uint16_t
```
Comment (Panquesito7) [inline searching/exponential_search.c:7]:
```suggestion
#include <assert.h>      /// for assert
#include <inttypes.h>    /// for int64_t, uint16_t
```

File: searching/exponential_search.c [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+    if ( upper_bound > length ) { upper_bound = length; }
+    // apply the binary search in the range
+    return binary_search(arr, lower_bound, upper_bound, n);
+}
+
+/**
+ *  Function: binary_search
+ *  \details recursive function that search the given element in
+ *  the array using the Binary Search algorithm
+ *  [Binary Search](https://github.com/TheAlgorithms/Algorithms-Explanation/blob/master/en/Search%20Algorithms/Binary%20Search.md)
```
Comment (Panquesito7) [inline searching/exponential_search.c:42]:
```suggestion
 *  the array using the [Binary Search](https://github.com/TheAlgorithms/Algorithms-Explanation/blob/master/en/Search%20Algorithms/Binary%20Search.md) algorithm
```

File: searching/exponential_search.c [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+    if ( l_index > r_index ) { return -1; }
+    if ( arr[middle_index] == n ) { return middle_index; }
+    // recursion
+    if ( arr[middle_index] > n ) { return binary_search(arr, l_index, middle_index-1, n); } // left
+    return binary_search(arr, middle_index+1, r_index, n); // right
+}
+
+/**
+ *  @brief Main function
+ *  @returns 0 on exit
```
Comment (Panquesito7) [inline searching/exponential_search.c:66]:
```suggestion
 * @brief Main function
 * @returns 0 on exit
```

File: searching/exponential_search.c [hunk 24]
```diff
@@ -0,0 +1,104 @@
...
+ */
+int main() 
+{
+    test();  // run self-test implementations
+    return 0;
+}
+
+/**
+ *  @brief Self-test implementations
+ *  @returns void
```
Comment (Panquesito7) [inline searching/exponential_search.c:76]:
```suggestion
 * @brief Self-test implementations
 * @returns void
```

File: searching/exponential_search.c [hunk 25]
```diff
@@ -0,0 +1,104 @@
...
+ */
+#include <assert.h>  // for assert
+#include <inttypes.h>  // for int64_t, uint16_t
+
+#define ELEMENT -10
+
+int64_t binary_search(const int64_t* arr, const uint16_t l_index, const uint16_t r_index, const int64_t n);
+int64_t exponential_search(const int64_t* arr, const uint16_t length, const int64_t n);
+static void test();
+
```
Comment (Panquesito7) [inline searching/exponential_search.c:14]: Add a one-line description of what the functions do (this should be a short description; see the example below).
```c++
static void test();  ///< used to run the self-test implementations
```

File: searching/exponential_search.c [hunk 26]
```diff
@@ -0,0 +1,104 @@
...
+/**
+ *  Function: exponential_search
```
Comment (Panquesito7) [inline searching/exponential_search.c:16]: Add a brief description of what the function does (same in other parts of the code).

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
👍 Good work!

File: searching/exponential_search.c [hunk 27]
```diff
@@ -0,0 +1,105 @@
...
+    // base cases
+    if ( l_index > r_index ) { return -1; }
+    if ( arr[middle_index] == n ) { return middle_index; }
+    // recursion
+    if ( arr[middle_index] > n ) { return binary_search(arr, l_index, middle_index-1, n); } // left
+    return binary_search(arr, middle_index+1, r_index, n); // right
+}
+
+/**
+ * @brief Main function that run the test
```
Comment (Panquesito7) [inline searching/exponential_search.c:66]:
```suggestion
 * @brief Main function
```

File: searching/exponential_search.c [hunk 28]
```diff
@@ -0,0 +1,105 @@
...
+#include <inttypes.h>  /// for int64_t, uint16_t
+
+#define ELEMENT -10
+
+int64_t binary_search(const int64_t* arr, const uint16_t l_index, const uint16_t r_index, const int64_t n); ///< used to perform the binary search over the given array
+int64_t exponential_search(const int64_t* arr, const uint16_t length, const int64_t n); ///< used to perform the exponential search over the given array
+static void test(); ///< used to run the self-test implementations
+
+/**
+ *  Function: exponential_search
```
Comment (Panquesito7) [inline searching/exponential_search.c:16]: There's a detailed description, but there's no brief description. 🙂

File: searching/exponential_search.c [hunk 29]
```diff
@@ -0,0 +1,105 @@
...
+ *  \details recursive function that take an array and quickly find the range 
+ *  where to apply the binary search algorithm to find the given element
+ *  ----------------------------
+ *  \param arr array where search the element
+ *  \param length the total length of the given array (arr)
+ *  \param n element to find in the array (arr)
+ * 
+ *  \returns the index of the element (n) in the array (arr)
+ *  \returns -1 if the element wasn't found
+ */
```
Comment (Panquesito7) [inline searching/exponential_search.c:26]:
```suggestion
 * \details recursive function that take an array and quickly find the range 
 * where to apply the binary search algorithm to find the given element
 * ----------------------------
 * \param arr array where search the element
 * \param length the total length of the given array (arr)
 * \param n element to find in the array (arr)
 *
 * \returns the index of the element (n) in the array (arr)
 * \returns -1 if the element wasn't found
 */
```

File: searching/exponential_search.c [hunk 30]
```diff
@@ -0,0 +1,105 @@
...
+ *  \details recursive function that search the given element in
+ *  the array using the [Binary Search](https://github.com/TheAlgorithms/Algorithms-Explanation/blob/master/en/Search%20Algorithms/Binary%20Search.md)
+ *  ----------------------------
+ *  \param arr array where search the element
+ *  \param l_index start index of the array (arr) to apply the algorithm
+ *  \param r_index end index of the array (arr) to apply the algorithm
+ *  \param n element to find in the array (arr)
+ *
+ *  \returns the index of the element (n) in the array (arr)
+ *  \returns -1 if the n element wasn't found
```
Comment (Panquesito7) [inline searching/exponential_search.c:51]:
```suggestion
 * Function: binary_search
 * \details recursive function that search the given element in
 * the array using the [Binary Search](https://github.com/TheAlgorithms/Algorithms-Explanation/blob/master/en/Search%20Algorithms/Binary%20Search.md)
 * ----------------------------
 * \param arr array where search the element
 * \param l_index start index of the array (arr) to apply the algorithm
 * \param r_index end index of the array (arr) to apply the algorithm
 * \param n element to find in the array (arr)
 *
 * \returns the index of the element (n) in the array (arr)
 * \returns -1 if the n element wasn't found
```

**Review by Panquesito7 — APPROVED** [review]
Amazing work, I like how you've made the code so well. 👍
Code and documentation are pretty good and refined; LGTM. 😄 🎉

**Comment by faridevnz** [comment]
> Amazing work, I like how you've made the code so well. 👍
> 
> Code and documentation are pretty good and refined; LGTM. 😄 🎉
> 
> 

@Panquesito7 thanks a lot 😆 I will continue to contribute 🥳

**Comment by KirwinD** [comment]
@faridevnz - great!

**Comment by faridevnz** [comment]
> @faridevnz - great!

Thanks @KirwinD 😊

**Pull request merged**